File: p4runtime/ConstraintValidator.kt
package fourward.p4runtime

import fourward.constraints.v1.ConstraintRequest
import fourward.constraints.v1.ConstraintResponse
import fourward.constraints.v1.LoadP4InfoRequest
import fourward.constraints.v1.ValidateEntryRequest
import java.io.Closeable
import java.io.DataInputStream
import java.io.DataOutputStream
import java.nio.file.Path
import java.util.concurrent.TimeUnit
import p4.config.v1.P4InfoOuterClass.P4Info
import p4.v1.P4RuntimeOuterClass.TableEntry

/**
 * Client for the p4-constraints validator subprocess.
 *
 * Manages the constraint_validator process lifecycle and validates table entries against
 * `@entry_restriction` / `@action_restriction` annotations. The wire protocol is length-delimited
 * protobuf over stdin/stdout (same framing as the simulator).
 */
class ConstraintValidator
private constructor(
  private val process: Process,
  private val input: DataInputStream,
  private val output: DataOutputStream,
  /** Number of tables with `@entry_restriction` constraints. */
  val constrainedTables: Int,
  /** Number of actions with `@action_restriction` constraints. */
  val constrainedActions: Int,
) : Closeable {

  /**
   * Validates a table entry against loaded constraints.
   *
   * @return null if the entry satisfies all constraints, or a human-readable violation explanation.
   * @throws ConstraintValidatorException if the validator process returns an error.
   */
  @Synchronized
  fun validateEntry(entry: TableEntry): String? {
    val request =
      ConstraintRequest.newBuilder()
        .setValidateEntry(ValidateEntryRequest.newBuilder().setEntry(entry))
        .build()
    val response = call(request)
    if (response.hasError()) {
      throw ConstraintValidatorException(response.error.message)
    }
    val violation = response.validateEntry.violation
    return violation.ifEmpty { null }
  }

  private fun call(request: ConstraintRequest): ConstraintResponse {
    val bytes = request.toByteArray()
    output.writeInt(bytes.size)
    output.write(bytes)
    output.flush()
    val length = input.readInt()
    val respBytes = ByteArray(length)
    input.readFully(respBytes)
    return ConstraintResponse.parseFrom(respBytes)
  }

  override fun close() {
    output.close()
    process.destroy()
    if (!process.waitFor(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
      process.destroyForcibly().waitFor()
    }
    input.close()
  }

  companion object {
    private const val SHUTDOWN_TIMEOUT_SECONDS = 5L

    /**
     * Creates a ConstraintValidator by spawning the validator subprocess and loading the given
     * P4Info.
     *
     * Returns null if the P4Info has no constraint annotations (no subprocess needed).
     *
     * @throws ConstraintValidatorException if the subprocess fails to load the P4Info.
     */
    fun create(p4info: P4Info, validatorBinary: Path): ConstraintValidator? {
      val process =
        ProcessBuilder(validatorBinary.toString())
          .redirectError(ProcessBuilder.Redirect.INHERIT)
          .start()
      val input = DataInputStream(process.inputStream.buffered())
      val output = DataOutputStream(process.outputStream.buffered())

      // Send LoadP4Info request.
      val request =
        ConstraintRequest.newBuilder()
          .setLoadP4Info(LoadP4InfoRequest.newBuilder().setP4Info(p4info))
          .build()
      val bytes = request.toByteArray()
      output.writeInt(bytes.size)
      output.write(bytes)
      output.flush()

      val length = input.readInt()
      val respBytes = ByteArray(length)
      input.readFully(respBytes)
      val response = ConstraintResponse.parseFrom(respBytes)

      if (response.hasError()) {
        process.destroyForcibly().waitFor()
        throw ConstraintValidatorException(
          "Failed to load P4Info into constraint validator: ${response.error.message}"
        )
      }

      val loadResponse = response.loadP4Info
      if (loadResponse.constrainedTables == 0 && loadResponse.constrainedActions == 0) {
        // No constraints — shut down the subprocess and return null.
        output.close()
        process.destroy()
        process.waitFor(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS)
        return null
      }

      return ConstraintValidator(
        process,
        input,
        output,
        loadResponse.constrainedTables,
        loadResponse.constrainedActions,
      )
    }
  }
}

class ConstraintValidatorException(message: String) : RuntimeException(message)

File: p4runtime/constraint_validator.cpp
// Constraint validator subprocess for the 4ward P4Runtime server.
//
// Reads ConstraintRequest messages from stdin and writes ConstraintResponse
// messages to stdout, using 4-byte big-endian length-delimited framing
// (same as the simulator protocol).
//
// On LoadP4Info: parses P4Info into ConstraintInfo via p4-constraints.
// On ValidateEntry: checks a TableEntry against loaded constraints.

#include <cstdint>
#include <iostream>
#include <optional>
#include <string>

#include "p4_constraints/backend/constraint_info.h"
#include "p4_constraints/backend/interpreter.h"
#include "p4runtime/constraint_validator.pb.h"

namespace {

// Reads a 4-byte big-endian length prefix from stdin, then reads that many
// bytes and parses them as a ConstraintRequest. Returns nullopt on EOF.
std::optional<fourward::constraints::v1::ConstraintRequest> ReadRequest() {
  uint32_t length = 0;
  for (int i = 0; i < 4; ++i) {
    int byte = std::cin.get();
    if (byte == EOF) return std::nullopt;
    length = (length << 8) | static_cast<uint8_t>(byte);
  }
  std::string buffer(length, '\0');
  if (!std::cin.read(buffer.data(), length)) return std::nullopt;
  fourward::constraints::v1::ConstraintRequest request;
  if (!request.ParseFromString(buffer)) return std::nullopt;
  return request;
}

// Writes a ConstraintResponse to stdout with a 4-byte big-endian length prefix.
void WriteResponse(
    const fourward::constraints::v1::ConstraintResponse& response) {
  std::string bytes;
  response.SerializeToString(&bytes);
  uint32_t length = bytes.size();
  for (int i = 3; i >= 0; --i) {
    std::cout.put(static_cast<char>((length >> (i * 8)) & 0xFF));
  }
  std::cout.write(bytes.data(), bytes.size());
  std::cout.flush();
}

void WriteError(const std::string& message) {
  fourward::constraints::v1::ConstraintResponse response;
  response.mutable_error()->set_message(message);
  WriteResponse(response);
}

}  // namespace

int main() {
  std::cerr << "constraint_validator starting" << std::endl;

  std::optional<p4_constraints::ConstraintInfo> constraint_info;

  while (auto request = ReadRequest()) {
    if (request->has_load_p4info()) {
      auto result =
          p4_constraints::P4ToConstraintInfo(request->load_p4info().p4info());
      if (!result.ok()) {
        WriteError(std::string(result.status().message()));
        continue;
      }
      constraint_info = std::move(*result);

      // Count tables and actions that have constraints.
      int constrained_tables = 0;
      int constrained_actions = 0;
      for (const auto& [id, table] : constraint_info->table_info_by_id) {
        if (table.constraint.has_value()) ++constrained_tables;
      }
      for (const auto& [id, action] : constraint_info->action_info_by_id) {
        if (action.constraint.has_value()) ++constrained_actions;
      }

      fourward::constraints::v1::ConstraintResponse response;
      auto* load_response = response.mutable_load_p4info();
      load_response->set_constrained_tables(constrained_tables);
      load_response->set_constrained_actions(constrained_actions);
      WriteResponse(response);
    } else if (request->has_validate_entry()) {
      if (!constraint_info.has_value()) {
        WriteError("No P4Info loaded; send LoadP4Info first");
        continue;
      }
      auto reason = p4_constraints::ReasonEntryViolatesConstraint(
          request->validate_entry().entry(), *constraint_info);
      if (!reason.ok()) {
        WriteError(std::string(reason.status().message()));
        continue;
      }
      fourward::constraints::v1::ConstraintResponse response;
      response.mutable_validate_entry()->set_violation(*reason);
      WriteResponse(response);
    } else {
      WriteError("Unknown request type");
    }
  }

  std::cerr << "constraint_validator exiting" << std::endl;
  return 0;
}
